File: app/src/main/java/com/bijoysingh/quicknote/activities/CreateOrEditAdvancedNoteActivity.java
package com.bijoysingh.quicknote.activities;

import android.os.Bundle;
import android.os.Handler;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.ImageView;

import com.bijoysingh.quicknote.R;
import com.bijoysingh.quicknote.activities.sheets.ColorPickerBottomSheet;
import com.bijoysingh.quicknote.database.Note;
import com.bijoysingh.quicknote.formats.Format;
import com.bijoysingh.quicknote.formats.FormatType;
import com.bijoysingh.quicknote.recyclerview.FormatTextViewHolder;
import com.bijoysingh.quicknote.recyclerview.SimpleItemTouchHelper;
import com.bijoysingh.quicknote.utils.CircleDrawable;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Calendar;
import java.util.Collections;

import static android.view.View.GONE;
import static com.bijoysingh.quicknote.utils.NoteBuilderKt.copyNote;

public class CreateOrEditAdvancedNoteActivity extends ViewAdvancedNoteActivity {

  private boolean active = false;
  private int maxUid = 0;

  private ImageView text;
  private ImageView heading;
  private ImageView subHeading;
  private ImageView checkList;
  private ImageView quote;
  private ImageView code;

  private Note lastNoteInstance = null;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setTouchListener();
    startHandler();

    if (getIntent().getBooleanExtra(ThemedActivity.Companion.getKey(), false)) {
      setNightMode(true);
    }
    lastNoteInstance = copyNote(note);
  }

  @Override
  protected void setEditMode() {
    setEditMode(getEditModeValue());
  }

  @Override
  protected boolean getEditModeValue() {
    return true;
  }

  private void setTouchListener() {
    ItemTouchHelper.Callback callback = new SimpleItemTouchHelper(adapter);
    ItemTouchHelper touchHelper = new ItemTouchHelper(callback);
    touchHelper.attachToRecyclerView(formatsView);
  }

  @Override
  protected void setNote() {
    super.setNote();
    maxUid = formats.size() + 1;
    boolean isEmpty = formats.isEmpty();
    if (isEmpty || formats.get(0).formatType != FormatType.HEADING) {
      addEmptyItem(0, FormatType.HEADING);
    }
    if (isEmpty) {
      addDefaultItem();
    }
  }

  protected void addDefaultItem() {
    addEmptyItem(FormatType.TEXT);
  }

  @Override
  protected void setButtonToolbar() {
    text = (ImageView) findViewById(R.id.format_text);
    text.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.TEXT);
      }
    });

    heading = (ImageView) findViewById(R.id.format_heading);
    heading.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.HEADING);
      }
    });

    subHeading = (ImageView) findViewById(R.id.format_sub_heading);
    subHeading.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.SUB_HEADING);
      }
    });

    checkList = (ImageView) findViewById(R.id.format_check_list);
    checkList.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.CHECKLIST_UNCHECKED);
      }
    });

    quote = (ImageView) findViewById(R.id.format_quote);
    quote.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.QUOTE);
      }
    });

    code = (ImageView) findViewById(R.id.format_code);
    code.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        addEmptyItemAtFocused(FormatType.CODE);
      }
    });
  }

  @Override
  protected void setTopToolbar() {
    actionDelete.setVisibility(GONE);
    actionShare.setVisibility(GONE);
    actionCopy.setVisibility(GONE);

    View colorButtonClicker = findViewById(R.id.color_button_clicker);
    colorButtonClicker.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        ColorPickerBottomSheet.Companion.openSheet(
            CreateOrEditAdvancedNoteActivity.this,
            new ColorPickerBottomSheet.ColorPickerController() {
              @Override
              public void onColorSelected(@NotNull Note note, int color) {
                setNoteColor(color);
              }

              @NotNull
              @Override
              public Note getNote() {
                return note;
              }
            });
      }
    });
  }

  @Override
  protected void notifyToolbarColor() {
    super.notifyToolbarColor();

    int toolbarIconColor = ContextCompat.getColor(
        context, isNightMode() ? R.color.white : R.color.material_blue_grey_700);
    text.setColorFilter(toolbarIconColor);
    heading.setColorFilter(toolbarIconColor);
    subHeading.setColorFilter(toolbarIconColor);
    checkList.setColorFilter(toolbarIconColor);
    quote.setColorFilter(toolbarIconColor);
    code.setColorFilter(toolbarIconColor);

    toolbar.setBackgroundColor(getColor(R.color.material_grey_50, R.color.material_grey_850));
  }

  @Override
  protected void onPause() {
    super.onPause();
    active = false;
    maybeUpdateNoteWithSync();
    destroyIfNeeded();
  }

  @Override
  public void onBackPressed() {
    super.onBackPressed();
    tryClosingTheKeyboard();
  }

  @Override
  protected void onResume() {
    super.onResume();
    active = true;
  }

  @Override
  protected void onResumeAction() {
    // do nothing
  }

  private boolean destroyIfNeeded() {
    if (note.isUnsaved()) {
      return true;
    }
    if (note.getFormats().isEmpty()) {
      note.delete(this);
      return true;
    }
    return false;
  }

  protected void maybeUpdateNoteWithoutSync() {
    maybeUpdateNote(false);
  }

  protected void maybeUpdateNoteWithSync() {
    maybeUpdateNote(true);
  }

  protected void maybeUpdateNote(boolean sync) {
    note.updateTimestamp = Calendar.getInstance().getTimeInMillis();
    note.description = Format.getNote(formats);

    // Ignore update if nothing changed. It allows for one undo per few seconds
    if (note.isEqual(lastNoteInstance)) {
      return;
    }

    maybeSaveNote(sync);
    lastNoteInstance.copyNote(note);
  }

  private void startHandler() {
    final Handler handler = new Handler();
    handler.postDelayed(new Runnable() {
      @Override
      public void run() {
        if (active) {
          maybeUpdateNoteWithoutSync();
          handler.postDelayed(this, HANDLER_UPDATE_TIME);
        }
      }
    }, HANDLER_UPDATE_TIME);
  }


  protected void addEmptyItem(FormatType type) {
    addEmptyItem(formats.size(), type);
  }

  private void addEmptyItem(int position, FormatType type) {
    Format format = new Format(type);
    format.uid = maxUid + 1;
    maxUid++;

    formats.add(position, format);
    adapter.addItem(format, position);
  }

  private void addEmptyItemAtFocused(FormatType type) {
    if (focusedFormat == null) {
      addEmptyItem(type);
      return;
    }

    int position = getFormatIndex(focusedFormat);
    if (position == -1) {
      addEmptyItem(type);
      return;
    }

    int newPosition = position + 1;
    addEmptyItem(newPosition, type);
    formatsView.getLayoutManager().scrollToPosition(newPosition);
    focus(newPosition);
  }

  public void focus(final int position) {
    Handler handler = new Handler();
    handler.postDelayed(new Runnable() {
      @Override
      public void run() {
        FormatTextViewHolder holder = findViewHolderAtPositionAggressively(position);
        if (holder == null) {
          return;
        }

        holder.requestEditTextFocus();
      }
    }, 100);
  }

  @Nullable
  private FormatTextViewHolder findViewHolderAtPositionAggressively(int position) {
    RecyclerView.ViewHolder holder = formatsView.findViewHolderForAdapterPosition(position);
    if (holder == null) {
      holder = formatsView.findViewHolderForLayoutPosition(position);
      if (holder == null) {
        return null;
      }
    }

    if (!(holder instanceof FormatTextViewHolder)) {
      return null;
    }

    return (FormatTextViewHolder) holder;
  }

  @Override
  protected void setNoteColor(int color) {
    note.color = color;
    colorButton.setBackground(new CircleDrawable(note.color));
  }

  @Override
  public void setFormat(Format format) {
    int position = getFormatIndex(format);
    if (position == -1) {
      return;
    }
    formats.set(position, format);
  }

  @Override
  public void moveFormat(int fromPosition, int toPosition) {
    if (fromPosition < toPosition) {
      for (int i = fromPosition; i < toPosition; i++) {
        Collections.swap(formats, i, i + 1);
      }
    } else {
      for (int i = fromPosition; i > toPosition; i--) {
        Collections.swap(formats, i, i - 1);
      }
    }
    maybeUpdateNoteWithoutSync();
  }

  @Override
  public void deleteFormat(Format format) {
    int position = getFormatIndex(format);
    if (position <= 0) {
      return;
    }
    focusedFormat = focusedFormat == null || focusedFormat.uid == format.uid ? null : focusedFormat;
    formats.remove(position);
    adapter.removeItem(position);
    maybeUpdateNoteWithoutSync();
  }

  @Override
  public void setFormatChecked(Format format, boolean checked) {
    // do nothing
  }

  @Override
  public void createOrChangeToNextFormat(Format format) {
    int position = getFormatIndex(format);
    if (position == -1) {
      return;
    }
    int newPosition = position + 1;
    if (newPosition < formats.size()) {
      focus(position + 1);
    } else {
      addEmptyItemAtFocused(Format.getNextFormatType(format.formatType));
    }
  }
}

File: app/src/main/java/com/bijoysingh/quicknote/activities/CreateOrEditAdvancedNoteActivity.kt
package com.bijoysingh.quicknote.activities

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.ImageView
import com.bijoysingh.quicknote.MaterialNotes.Companion.appTheme
import com.bijoysingh.quicknote.R
import com.bijoysingh.quicknote.activities.sheets.ColorPickerBottomSheet
import com.bijoysingh.quicknote.activities.sheets.NoteFormatOptionsBottomSheet
import com.bijoysingh.quicknote.activities.sheets.NoteMarkdownOptionsBottomSheet
import com.bijoysingh.quicknote.database.utils.*
import com.bijoysingh.quicknote.formats.recycler.FormatImageViewHolder
import com.bijoysingh.quicknote.formats.recycler.FormatTextViewHolder
import com.bijoysingh.quicknote.recyclerview.SimpleItemTouchHelper
import com.bijoysingh.quicknote.utils.CircleDrawable
import com.bijoysingh.quicknote.utils.ThemeColorType
import com.bijoysingh.quicknote.utils.bind
import com.maubis.scarlet.base.database.room.note.Note
import com.maubis.scarlet.base.format.*
import com.maubis.scarlet.base.note.NoteBuilder
import com.maubis.scarlet.base.note.NoteImage
import pl.aprilapps.easyphotopicker.DefaultCallback
import pl.aprilapps.easyphotopicker.EasyImage
import java.io.File
import java.util.*


open class CreateOrEditAdvancedNoteActivity : ViewAdvancedNoteActivity() {

  private var active = false
  private var lastNoteInstance: Note? = null
  private var maxUid = 0
  private var toolbarMode: ToolbarMode = ToolbarMode.FORMAT

  val text: ImageView by bind(R.id.format_text)
  val heading: ImageView by bind(R.id.format_heading)
  val subHeading: ImageView by bind(R.id.format_sub_heading)
  val checkList: ImageView by bind(R.id.format_check_list)
  val formatMore: ImageView by bind(R.id.format_more)

  val markdownBold: ImageView by bind(R.id.markdown_bold)
  val markdownHeading: ImageView by bind(R.id.markdown_heading)
  val markdownItalics: ImageView by bind(R.id.markdown_italics)
  val markdownUnordered: ImageView by bind(R.id.markdown_unordered)
  val markdownMore: ImageView by bind(R.id.markdown_more)

  val chevronLeft: ImageView by bind(R.id.toolbar_chevron_left)
  val chevronRight: ImageView by bind(R.id.toolbar_chevron_right)

  override val editModeValue: Boolean get() = true

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setTouchListener()
    startHandler()
    lastNoteInstance = NoteBuilder().copy(note!!)
  }

  override fun setEditMode() {
    setEditMode(editModeValue)
  }

  private fun setTouchListener() {
    val callback = SimpleItemTouchHelper(adapter)
    val touchHelper = ItemTouchHelper(callback)
    touchHelper.attachToRecyclerView(formatsView)
  }

  override fun setNote() {
    super.setNote()
    maxUid = formats.size + 1
    val isEmpty = formats.isEmpty()
    if (isEmpty || (formats[0].formatType !== FormatType.HEADING
            && formats[0].formatType !== FormatType.IMAGE)) {
      addEmptyItem(0, FormatType.HEADING)
    }
    if (isEmpty) {
      addDefaultItem()
    }
  }

  protected open fun addDefaultItem() {
    addEmptyItem(FormatType.TEXT)
  }

  override fun setBottomToolbar() {
    setFormatToolbar()
    setMarkdownButtonToolbar()
    chevronLeft.setOnClickListener { toggleToolbarMode() }
    chevronRight.setOnClickListener { toggleToolbarMode() }
  }

  fun setFormatToolbar() {
    text.setOnClickListener { addEmptyItemAtFocused(FormatType.TEXT) }
    heading.setOnClickListener { addEmptyItemAtFocused(FormatType.HEADING) }
    subHeading.setOnClickListener { addEmptyItemAtFocused(FormatType.SUB_HEADING) }
    checkList.setOnClickListener { addEmptyItemAtFocused(FormatType.CHECKLIST_UNCHECKED) }
    formatMore.setOnClickListener {
      NoteFormatOptionsBottomSheet.openSheet(this)
    }
  }

  fun setMarkdownButtonToolbar() {
    markdownBold.setOnClickListener { triggerMarkdown(MarkdownType.BOLD) }
    markdownItalics.setOnClickListener { triggerMarkdown(MarkdownType.ITALICS) }
    markdownHeading.setOnClickListener { triggerMarkdown(MarkdownType.HEADER) }
    markdownUnordered.setOnClickListener { triggerMarkdown(MarkdownType.UNORDERED) }
    markdownMore.setOnClickListener {
      NoteMarkdownOptionsBottomSheet.openSheet(this)
    }
  }

  fun toggleToolbarMode() {
    toolbarMode = when (toolbarMode) {
      ToolbarMode.FORMAT -> ToolbarMode.MARKDOWN
      ToolbarMode.MARKDOWN -> ToolbarMode.FORMAT
    }
    notifyToolbarModeChange()
  }

  fun notifyToolbarModeChange() {
    when (toolbarMode) {
      ToolbarMode.FORMAT -> {
        formatToolbar.visibility = VISIBLE
        markdownToolbar.visibility = GONE
      }
      ToolbarMode.MARKDOWN -> {
        formatToolbar.visibility = GONE
        markdownToolbar.visibility = VISIBLE
      }
    }
  }

  override fun setTopToolbar() {
    actionDelete.visibility = GONE
    actionShare.visibility = GONE
    actionCopy.visibility = GONE

    val colorButtonClicker = findViewById<View>(R.id.color_button_clicker)
    colorButtonClicker.setOnClickListener {
      ColorPickerBottomSheet.openSheet(
          this@CreateOrEditAdvancedNoteActivity,
          object : ColorPickerBottomSheet.ColorPickerController {
            override fun onColorSelected(note: Note, color: Int) = setNoteColor(color)
            override fun getNote(): Note = note!!
          })
    }
  }

  override fun notifyToolbarColor() {
    super.notifyToolbarColor()
    val theme = appTheme()
    toolbar.setBackgroundColor(theme.get(ThemeColorType.TOOLBAR_BACKGROUND))
    markdownToolbar.setBackgroundColor(theme.get(ThemeColorType.TOOLBAR_BACKGROUND))

    val toolbarIconColor = theme.get(ThemeColorType.TOOLBAR_ICON)
    text.setColorFilter(toolbarIconColor)
    heading.setColorFilter(toolbarIconColor)
    subHeading.setColorFilter(toolbarIconColor)
    checkList.setColorFilter(toolbarIconColor)
    formatMore.setColorFilter(toolbarIconColor)

    markdownHeading.setColorFilter(toolbarIconColor)
    markdownBold.setColorFilter(toolbarIconColor)
    markdownUnordered.setColorFilter(toolbarIconColor)
    markdownItalics.setColorFilter(toolbarIconColor)
    markdownMore.setColorFilter(toolbarIconColor)

    val hintColor = theme.get(ThemeColorType.PRIMARY_TEXT)
    chevronLeft.setColorFilter(hintColor)
    chevronRight.setColorFilter(hintColor)
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    EasyImage.handleActivityResult(requestCode, resultCode, data, this, object : DefaultCallback() {
      override fun onImagePicked(imageFile: File?, source: EasyImage.ImageSource?, type: Int) {
        if (imageFile == null) {
          return
        }

        val targetFile = NoteImage(context).renameOrCopy(note!!, imageFile)
        val index = getFormatIndex(type)
        triggerImageLoaded(index, targetFile)
      }

      override fun onImagePickerError(e: Exception, source: EasyImage.ImageSource, type: Int) {
        //Some error handling
      }
    })
  }

  override fun onPause() {
    super.onPause()
    active = false
    maybeUpdateNoteWithoutSync()
    val destroyed = destroyIfNeeded()
    if (!destroyed) {
      note!!.saveToSync()
    }
  }

  override fun onBackPressed() {
    super.onBackPressed()
    tryClosingTheKeyboard()
  }

  override fun onResume() {
    super.onResume()
    active = true
  }

  override fun onResumeAction() {
    // do nothing
  }

  private fun destroyIfNeeded(): Boolean {
    if (note!!.isUnsaved()) {
      return true
    }
    if (note!!.getFormats().isEmpty()) {
      note!!.delete(this)
      return true
    }
    return false
  }

  protected fun maybeUpdateNoteWithoutSync() {
    val vNote = note!!
    val vLastNoteInstance = lastNoteInstance ?: note!!

    vNote.description = FormatBuilder().getDescription(formats)

    // Ignore update if nothing changed. It allows for one undo per few seconds
    if (vNote.isEqual(vLastNoteInstance)) {
      return
    }

    vNote.updateTimestamp = Calendar.getInstance().timeInMillis
    maybeSaveNote(false)
    vLastNoteInstance.copyNote(vNote)
  }

  private fun startHandler() {
    val handler = Handler()
    handler.postDelayed(object : Runnable {
      override fun run() {
        if (active) {
          maybeUpdateNoteWithoutSync()
          handler.postDelayed(this, ViewAdvancedNoteActivity.Companion.HANDLER_UPDATE_TIME.toLong())
        }
      }
    }, ViewAdvancedNoteActivity.Companion.HANDLER_UPDATE_TIME.toLong())
  }

  protected fun addEmptyItem(type: FormatType) {
    addEmptyItem(formats.size, type)
  }

  private fun addEmptyItem(position: Int, type: FormatType) {
    val format = Format(type)
    format.uid = maxUid + 1
    maxUid++

    formats.add(position, format)
    adapter.addItem(format, position)
  }

  fun addEmptyItemAtFocused(type: FormatType) {
    if (focusedFormat == null) {
      addEmptyItem(type)
      return
    }

    val position = getFormatIndex(focusedFormat!!)
    if (position == -1) {
      addEmptyItem(type)
      return
    }

    val newPosition = position + 1
    addEmptyItem(newPosition, type)
    formatsView.layoutManager.scrollToPosition(newPosition)
    focus(newPosition)
  }

  fun focus(position: Int) {
    val handler = Handler()
    handler.postDelayed(Runnable {
      val holder = findTextViewHolderAtPosition(position) ?: return@Runnable
      holder.requestEditTextFocus()
    }, 100)
  }

  fun triggerMarkdown(markdownType: MarkdownType) {
    if (focusedFormat == null) {
      return
    }

    val position = getFormatIndex(focusedFormat!!)
    if (position == -1) {
      return
    }

    val handler = Handler()
    handler.postDelayed(Runnable {
      val holder = findTextViewHolderAtPosition(position) ?: return@Runnable
      holder.requestMarkdownAction(markdownType)
    }, 100)
  }

  fun triggerImageLoaded(position: Int, file: File) {
    if (position == -1) {
      return
    }

    val holder = findImageViewHolderAtPosition(position) ?: return
    holder.populateFile(file)

    val formatToChange = formats[position]
    if (!formatToChange.text.isBlank()) {
      val noteImage = NoteImage(context)
      noteImage.deleteIfExist(noteImage.getFile(note!!.uuid, formatToChange.text))
    }
    formatToChange.text = file.name
    setFormat(formatToChange)
  }

  private fun findTextViewHolderAtPosition(position: Int): FormatTextViewHolder? {
    val holder = findViewHolderAtPositionAggressively(position)
    return if (holder !== null && holder is FormatTextViewHolder) holder else null
  }

  private fun findImageViewHolderAtPosition(position: Int): FormatImageViewHolder? {
    val holder = findViewHolderAtPositionAggressively(position)
    return if (holder !== null && holder is FormatImageViewHolder) holder else null
  }

  private fun findViewHolderAtPositionAggressively(position: Int): RecyclerView.ViewHolder? {
    var holder: RecyclerView.ViewHolder? = formatsView.findViewHolderForAdapterPosition(position)
    if (holder == null) {
      holder = formatsView.findViewHolderForLayoutPosition(position)
      if (holder == null) {
        return null
      }
    }
    return holder
  }

  override fun setNoteColor(color: Int) {
    note!!.color = color
    colorButton.background = CircleDrawable(note!!.color)
  }

  override fun setFormat(format: Format) {
    val position = getFormatIndex(format)
    if (position == -1) {
      return
    }
    formats[position] = format
  }

  override fun moveFormat(fromPosition: Int, toPosition: Int) {
    if (fromPosition < toPosition) {
      for (i in fromPosition until toPosition) {
        Collections.swap(formats, i, i + 1)
      }
    } else {
      for (i in fromPosition downTo toPosition + 1) {
        Collections.swap(formats, i, i - 1)
      }
    }
    maybeUpdateNoteWithoutSync()
  }

  override fun deleteFormat(format: Format) {
    val position = getFormatIndex(format)
    if (position < 0) {
      return
    }
    focusedFormat = if (focusedFormat == null || focusedFormat!!.uid == format.uid) null else focusedFormat
    formats.removeAt(position)
    adapter.removeItem(position)
    maybeUpdateNoteWithoutSync()
  }

  override fun setFormatChecked(format: Format, checked: Boolean) {
    // do nothing
  }

  override fun createOrChangeToNextFormat(format: Format) {
    val position = getFormatIndex(format)
    if (position == -1) {
      return
    }

    val isCheckList = (format.formatType === FormatType.CHECKLIST_UNCHECKED
        || format.formatType === FormatType.CHECKLIST_CHECKED)
    val newPosition = position + 1
    when {
      isCheckList -> addEmptyItemAtFocused(FormatBuilder().getNextFormatType(FormatType.CHECKLIST_UNCHECKED))
      newPosition < formats.size -> focus(position + 1)
      else -> addEmptyItemAtFocused(FormatBuilder().getNextFormatType(format.formatType))
    }
  }
}
